### PR TITLE
Fix bug in cpp iteration over rvalue-dependant attribute

### PR DESCRIPTION
**Pull request by xavth**
The implementation of the Python-style `for ... in` loop over an iterable cpp sequence (using `begin()` and `end()`) needs to determine whether the sequence should be stored in a temporary variable or if it can be iterated over directly.

The previous code assumed that attributes are always safe to be iterated directly, but attributes can depend on an rvalue, as in:

    for i in returns_object_with_vector_attribute().vector:
        ...

In such a case we need to store the `in` operand in a temporary value on which `begin()` and `end()` can be safely called.

It might make sense to move the test into tests/run/cpp_iterators

### CODE REVIEW

**Comment by xavth** [comment]
Should I open an issue ? Should I copy the commit message to the PR description ?

**Comment by da-woods** [comment]
> Should I open an issue ? Should I copy the commit message to the PR description ?

Definitely copy the commit message here so that the PR makes sense. I'm not convinced that an issue is _really_ needed but others might disagree.

**Comment by xavth** [comment]
> > Should I open an issue ? Should I copy the commit message to the PR description ?
> 
> Definitely copy the commit message here so that the PR makes sense. I'm not convinced that an issue is _really_ needed but others might disagree.

OK, I updated the PR description with the description in the commit message.

**Comment by xavth** [comment]
I noticed that the test builds on OSX seemed to fail because I used C++11-style list-initialization in the external C++ code to support my testcase, so I modified it.

**Comment by da-woods** [comment]
Just closing and re-opening to re-run the tests (just trying to see if I can clear up some old inactive PRs)

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -2937,6 +2937,33 @@ def generate_iter_next_result_code(self, result_name, code):
...
+            # postponed from CppIteratorNode.generate_subexpr_disposal_code
+            # and CppIteratorNode.free_subexpr_temps
+            ExprNode.generate_subexpr_disposal_code(self, code)
+            ExprNode.free_subexpr_temps(self, code)
+            # adapt relevant bits from ExprNodes.generate_disposal_code
+            if self.has_temp_moved:
+                code.globalstate.use_utility_code(
+                    UtilityCode.load_cached("MoveIfSupported", "CppSupport.cpp"))
+        else:
+            ExprNode.generate_disposal_code(self, code)
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:2965]:
```suggestion
                    ExprNode.generate_disposal_code(self, code)
```

possibly wrong but I'd like to try it....

**Comment by da-woods** [comment]
I've moved the tests into the main cpp_iterators file (but made no other changes to them).

I think I'm happy with this PR and would like to merge it. I'll wait for a bit to see if there's other comments/objections though

**Comment by da-woods** [comment]
I'd been planning to merge this after the next release - I think it's a worthwhile (if slightly obscure) bug fix (link back to https://github.com/cython/cython/issues/4022)

**Comment by da-woods** [comment]
Thanks, and sorry it took a while to merge

**Comment by xavth** [comment]
Thanks for merging @da-woods :) and sorry in turn I didn't answer sooner.

**Pull request merged**